File: solutions/typescript/problems/[0547_Medium] Number of Provinces.ts
// @ts-ignore
class DisjointSet {
    private root: number[] = [];
    private rank: number[] = [];

    constructor(size: number) {
        for (let i = 0; i < size; i++) {
            this.root[i] = i;
            this.rank[i] = 1;
        }
    }

    find(x: number): number {
        if (x === this.root[x]) {
            return x;
        }
        return (this.root[x] = this.find(this.root[x]));
    }

    union(x: number, y: number): void {
        const rootX = this.find(x);
        const rootY = this.find(y);
        if (rootX !== rootY) {
            if (this.rank[x] > this.rank[y]) {
                this.root[rootY] = rootX;
            } else if (this.rank[x] < this.rank[y]) {
                this.root[rootX] = rootY;
            } else {
                this.root[rootY] = rootX;
                this.rank[rootX]++;
            }
        }
    }

    connected(x: number, y: number): boolean {
        return this.find(x) === this.find(y);
    }
}

function findCircleNum(isConnected: number[][]): number {
    const disjointSet = new DisjointSet(isConnected.length);
    for (let i = 0; i < isConnected.length; i++) {
        for (let j = i; j < isConnected[i].length; j++) {
            if (isConnected[i][j]) {
                disjointSet.union(i, j);
            }
        }
    }
    const provinces = new Set();
    for (let i = 0; i < isConnected.length; i++) {
        provinces.add(disjointSet.find(i));
    }
    return provinces.size;
}

File: solutions/c++/problems/[0547_Medium] Number of Provinces.cpp
class DisjointSet {
    vector<int> root;
    vector<int> rank;
public:
    DisjointSet(int n) {
        root.reserve(n);
        rank = vector<int>(n, 0);
        for (int i = 0; i < n; i++) root[i] = i;
    }

    void Union(int x, int y) {
        int rootX = Find(x);
        int rootY = Find(y);
        if (rank[rootX] < rank[rootY]) {
            root[rootX] = rootY;
        } else {
            root[rootY] = rootX;
            if (rank[rootX] == rank[rootY]) rank[rootX]++;
        }
    }

    int Find(int x) {
        if (root[x] == x) return x;
        return root[x] = root[root[x]];
    }

    bool connected(int x, int y) {
        return Find(x) == Find(y);
    }
};

class Solution {
public:
    int findCircleNum(vector <vector<int>> &isConnected) {
        int n = isConnected.size();
        auto ds = new DisjointSet(n);
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isConnected[i][j] == 1) ds->Union(i, j);
            }
        }
        vector<int> provinces;
        for (int i = 0; i < n; i++) {
            bool shouldAdd = true;
            for (int root: provinces) {
                if (ds->connected(i, root)) {
                    shouldAdd = false;
                    break;
                }
            }
            if (shouldAdd) provinces.push_back(i);
        }
        return provinces.size();
    }
};